vapi/library/library_item_updatesession_file.go

```diff
@@ -0,0 +1,126 @@
+	Name           string          `json:"name,omitempty"`
+	Size           *int64          `json:"size,omitempty"`
+	SourceEndpoint *SourceEndpoint `json:"source_endpoint,omitempty"`
+	SourceType     string          `json:"source_type,omitempty"`
+}
+
+// UpdateFileInfo is the response specification for the updatesession
+// operations file:add, file:get, and file:list.
+type UpdateFileInfo struct {
+	Name             string         `json:"name"`
+	SourceType       string         `json:"source_type"`
+	Status           string         `json:"status"`
+	BytesTransferred int64          `json:"bytes_transferred"`
+	Size             int64          `json:"size"`
+	ChecksumInfo     Checksum       `json:"checksum_info"`
+	SourceEndpoint   SourceEndpoint `json:"source_endpoint"`
+	UploadEndpoint   SourceEndpoint `json:"upload_endpoint"`
+}
+
+// AddLibraryItemFile adds a file
+func (c *Manager) AddLibraryItemFile(ctx context.Context, sessionID string, updateFile UpdateFile) (*UpdateFileInfo, error) {
+	url := internal.URL(c, internal.LibraryItemUpdateSessionFile).WithID(sessionID).WithAction("add")
+	spec := struct {
+		FileSpec UpdateFile `json:"file_spec"`
+	}{updateFile}
+	var res UpdateFileInfo
+	return &res, c.Do(ctx, url.Request(http.MethodPost, spec), &res)
+}
+
+// AddLibraryItemFileFromURI adds a file from a remote URI.
+func (c *Manager) AddLibraryItemFileFromURI(
+	ctx context.Context,
+	sessionID, fileName, uri string) (*UpdateFileInfo, error) {
+
+	n, fingerprint, err := GetContentLengthAndFingerprint(ctx, uri)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := c.AddLibraryItemFile(ctx, sessionID, UpdateFile{
+		Name:       fileName,
+		SourceType: "PULL",
+		Size:       &n,
+		SourceEndpoint: &SourceEndpoint{
+			URI:                      uri,
+			SSLCertificateThumbprint: fingerprint,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return info, c.CompleteLibraryItemUpdateSession(ctx, sessionID)
+}
+
+// GetLibraryItemUpdateSessionFile retrieves information about a specific file
+// that is a part of an update session.
+func (c *Manager) GetLibraryItemUpdateSessionFile(ctx context.Context, sessionID string, fileName string) (*UpdateFileInfo, error) {
+	url := internal.URL(c, internal.LibraryItemUpdateSessionFile).WithID(sessionID).WithAction("get")
+	spec := struct {
+		Name string `json:"file_name"`
+	}{fileName}
+	var res UpdateFileInfo
+	return &res, c.Do(ctx, url.Request(http.MethodPost, spec), &res)
+}
+
+// GetContentLengthAndFingerprint gets the number of bytes returned
+// by the URI as well as the SHA1 fingerprint of the peer certificate
+// if the URI's scheme is https.
+func GetContentLengthAndFingerprint(
+	ctx context.Context, uri string) (int64, string, error) {
+	resp, err := http.Head(uri)
+	if err != nil {
+		return 0, "", err
+	}
+	if resp.TLS == nil || len(resp.TLS.PeerCertificates) == 0 {
+		return resp.ContentLength, "", nil
+	}
+	fingerprint := &bytes.Buffer{}
```

Comment:
Can use `soap.ThumbprintSHA1` here.

Reply:
Hi @dougm,

I suppose, but I don't think it's a big deal either way.